pennylane/math/_is_independent.py

```diff
@@ -0,0 +1,317 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains the _is_independent function that checks
+a function to be independent of its arguments for the interfaces
+
+* Autograd
+* JAX
+* TensorFlow
+* PyTorch
+"""
+import warnings
+
+import numpy as np
+
+from autograd.tracer import isbox, new_box, trace_stack
+from autograd.core import VJPNode
+
+
+def _autograd_is_independent_ana(func, *args, **kwargs):
+    """Test analytically whether a function is independent of its arguments
+    using Autograd.
+
+    Args:
+        func (callable): Function to test for independence
+        args (tuple): Arguments for the function with respect to which
+            to test for independence
+        kwargs (dict): Keyword arguments for the function at which
+            (but not with respect to which) to test for independence
+
+    Returns:
+        bool: Whether the function seems to not depend on it ``args``
+        analytically. That is, an output of ``True`` means that the
+        ``args`` do *not* feed into the output.
+
+    In Autograd, we test this by sending a ``Box`` through the function and
+    testing whether the output is again a ``Box`` and on the same trace as
+    the input ``Box``. This means that we can trace actual *independence*
+    of the output from the input, not only whether the passed function is
+    constant.
+    The code is adapted from
+    `autograd.tracer.py::trace
+    <https://github.com/HIPS/autograd/blob/master/autograd/tracer.py#L7>`__.
+    """
+    # pylint: disable=protected-access
+    node = VJPNode.new_root()
+    with trace_stack.new_trace() as t:
+        start_box = new_box(args, t, node)
+        end_box = func(*start_box, **kwargs)
+
+    if type(end_box) in [tuple, list]:
```

Comment:
Yes, true!
On a tangent: In some `autograd` source code I saw that `type(x) is y` is 3 times faster than `isinstance(x, y)`.
However, this makes sense because getting the `type` of `x` or the `type` and all parent types are two quite different things..

Reply:
oh I didn't know that - if `type` is faster, lets leave it as that